### PR TITLE
Implement dotnet test help option

### PR DESCRIPTION
**Pull request by mariam-abdulla**

### CODE REVIEW

**Review by MarcoRossignoli — COMMENTED** [review]
Did review round some question and cleanup

File: sdk.sln
Comment (mariam-abdulla) [reply sdk.sln]: it's not needed. Updated.

File: src/Cli/dotnet/commands/dotnet-test/IPC/Models/Module.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/IPC/Models/Module.cs]: Added a comment in RegisterSerializer.cs

File: src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs]: Added.

File: src/Cli/dotnet/commands/dotnet-test/IPC/Serializers/VoidResponseSerializer.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/IPC/Serializers/VoidResponseSerializer.cs]: Not really.

File: src/Cli/dotnet/commands/dotnet-test/TestApplication.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestApplication.cs]: Updated.

File: src/Cli/dotnet/commands/dotnet-test/TestApplication.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestApplication.cs]: Updated.

File: src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs]: Updated.

File: src/Cli/dotnet/Program.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/Program.cs]: Done.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs]: Updated.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs]: Updated.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:41]: Fixed.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs]: Updated.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs]: Fixed.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs]: Updated.

File: src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs
Comment (baronfel) [inline src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs:26]: Right now there are going to be quite a few instances of this pipe client created over the course of a build:

* one per outer build
* one per inner build (aka per-TFM)

We suggest instead that you use the MSBuild Engine's [RegisterTaskObject API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.build.framework.ibuildengine4.registertaskobject?view=msbuild-17-netcore) to create one client per `GetTestsProjectPipeName` and re-use it. If you use this API and your type also implements `IDisposable`, then MSBuild will also make sure that it is disposed correctly.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (baronfel) [inline src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:78]: Since this is a new IPC, is there a threat model/review that's been done?

**Review by baronfel — COMMENTED** [review]
A few notes based on discussing the PR with @rainersigwald earlier today.

File: src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestingPlatformCommand.cs:78]: Not at the moment

@drognanar for the json-rpc protocol have you done the threat model and review with someone?

File: src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs
Comment (MarcoRossignoli) [reply src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs:10]: Make sense thanks

File: src/Cli/dotnet/commands/dotnet-test/TestApplication.cs
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestApplication.cs:57]: Sure this is clearly incomplete we will add extensive testing and use cases.

File: src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs
Comment (MarcoRossignoli) [reply src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs:26]: This is interesting thanks

File: src/Cli/dotnet/commands/dotnet-test/IPC/INamedPipeIC.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/INamedPipeIC.cs:8]: Why generic when nothing is generic?

File: src/Cli/dotnet/commands/dotnet-test/IPC/INamedPipeIC.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/INamedPipeIC.cs:6]: Type doesn't match file name

File: src/Cli/dotnet/commands/dotnet-test/IPC/Models/CommandLineOptionMessages.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/Models/CommandLineOptionMessages.cs:8]: Nit: `record` are `class` by default

```suggestion
    internal sealed record CommandLineOptionMessage(string Name, string Description, bool IsHidden, bool IsBuiltIn) : IRequest;

    internal sealed record CommandLineOptionMessages(string ModuleName, CommandLineOptionMessage[] CommandLineOptionMessageList) : IRequest;
```

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs:17]: In such context, I think it's better to have an argument of type `Type`. It's never possible to get the compiler infer the type because it's not used as argument.

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs:26]: Maybe `ArgumentException` here?

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs:28]: You probably want to use resx to allow localization.

Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeBase.cs:28]: I would use `with ID`.

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs:57]: Do we need all these pragmas? Is `dotnet` compiled from something else that netcore?

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs:191]: Nit:
```suggestion

    public void Dispose()
```

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs:52]: As far as I can tell this can be made readonly

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs:76]: Nit

```suggestion
        _loopTask = Task.Run(
            async () =>
            {
                try
                {
                    await InternalLoopAsync(_cancellationToken);
                }
                catch (OperationCanceledException ex) when (ex.CancellationToken == _cancellationToken)
                {
                    // We are being cancelled, so we don't need to wait anymore
                    return;
                }
                catch (Exception ex)
                {
                    Environment.FailFast($"[NamedPipeServer] Unhandled exception:{Environment.NewLine}{ex}", ex);
                }
            }, cancellationToken);
```

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs:89]: It probably doesn't change much but might be worth to do 

```suggestion
        while (!cancellationToken.IsCancellationRequested)
```

File: src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs:4]: We are in the SDK, is this a copy-paste error?

File: src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs:10]: Class could be static.

File: src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs
Comment (Evangelink) [inline src/Cli/dotnet/commands/dotnet-test/IPC/ObjectFieldIds.cs:16]: Class could be static.

**Review by Evangelink — COMMENTED** [review]
A lot of error message strings should be extracted as resx for translation.

File: src/Cli/dotnet/commands/dotnet-test/IPC/INamedPipeIC.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/IPC/INamedPipeIC.cs:8]: This was copied from testfx repo.

File: src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs
Comment (mariam-abdulla) [reply src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeClient.cs:57]: Those files are included in the Microsoft.NET.Build.Extensions.Tasks (which uses both net47 and net9) project, because we have a task there that uses them.

File: src/Layout/redist/MSBuildImports/Current/Microsoft.Common.targets/ImportAfter/Microsoft.TestPlatform.ImportAfter.targets
Comment (MarcoRossignoli) [reply src/Layout/redist/MSBuildImports/Current/Microsoft.Common.targets/ImportAfter/Microsoft.TestPlatform.ImportAfter.targets:20]: Now we moved to the common tasks project so I suppose we can keep it as-is.

@baronfel can you give a quick review so we can merge this first step(is opted out but code start to be too much to review).

File: src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs
Comment (MarcoRossignoli) [reply src/Tasks/Microsoft.NET.Testing.Tasks/GetTestsProject.cs:10]: @mariam-abdulla we can remove this project now right?

**Review by MarcoRossignoli — COMMENTED** [review]
Some cleanup(remove custom task project) around but I suppose we can merge after it this first part cc @baronfel 

@mariam-abdulla ensure green PR.

File: src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs:185]: Yep it's possible to opt-in only in debug...you can see in line 170 in release we go with only old code, is it fine to you this "way to opt-out" or do you have historically different way to do incremental merge?

Comment (baronfel) [reply src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs:185]: This is totally fine - I just wanted to make sure the intent.  It lowers the 'bar' if the new code won't be accessible by users :)

Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/TestCommandParser.cs:185]: The env var is only for development use we will open a design doc to add our testing section inside the global.json where we will describe how to officially opt-in the sdk for the new testing platform keeping for now the old one as "default".

File: src/Cli/dotnet/commands/dotnet-test/IPC/IClient.cs
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-test/IPC/IClient.cs:9]: I think you won't have to disable the warnings if you reverse the order of these interfaces, i.e.,
IClient :
#if NETCOREAPP
IAsyncDisposable,
#endif
IDisposable

**Comment by Forgind** [comment]
There seems to be a lot of code (like implementations for IPC things) that looks like it was probably pulled from another repo (runtime?). Why do we need separate implementations here?

**Comment by MarcoRossignoli** [comment]
> here seems to be a lot of code (like implementations for IPC things) that looks like it was probably pulled from another repo (runtime?). Why do we need separate implementations here?

It's not from runtime is a custom binary protocol implementation used by the new testing platform to support special scenario, you can read the architecture and the needs for it here https://github.com/microsoft/testfx/blob/main/docs/testingplatform/architecture.md

So we're using the same IPC system of the new testing plat, the code is battle tested and coming from https://github.com/microsoft/testfx/tree/main/src/Platform/Microsoft.Testing.Platform/IPC

The new testing platform doesn't need any dependency so we don't export a nuget package to "consume it". We opted for "code move" to have the same basic framing protocol.

File: src/Cli/dotnet/commands/dotnet-test/IPC/IClient.cs
Comment (MarcoRossignoli) [reply src/Cli/dotnet/commands/dotnet-test/IPC/IClient.cs:9]: @mariam-abdulla pls give it a try looks a good idea! thanks

**Comment by mariam-abdulla** [comment]
@dotnet/domestic-cat We are using $(MicrosoftNETBuildTasksAssembly) in Microsoft.TestPlatform.ImportAfter.targets
but seems to be problematic in tests. Can you recommend what can be done in that case?

**Comment by Forgind** [comment]
> @dotnet/domestic-cat We are using $(MicrosoftNETBuildTasksAssembly) in Microsoft.TestPlatform.ImportAfter.targets but seems to be problematic in tests. Can you please recommend what can be done in that case?

That seems to be defined in Microsoft.NET.Sdk.Common.targets, which only comes in if you either import it directly or reference it indirectly by, for instance, using Sdk="Microsoft.NET.Sdk". I looked at one of the failing tests, and it was not an SDK-style project, and it didn't import that .targets file directly, which explains why it evaluated to "". You should be able to change the test asset to import it (and make sure you import it before Microsoft.TestPlatform.ImportAfter.targets).

Do you need this task in legacy .*proj files?

**Comment by mariam-abdulla** [comment]
> > @dotnet/domestic-cat We are using $(MicrosoftNETBuildTasksAssembly) in Microsoft.TestPlatform.ImportAfter.targets but seems to be problematic in tests. Can you please recommend what can be done in that case?
> 
> That seems to be defined in Microsoft.NET.Sdk.Common.targets, which only comes in if you either import it directly or reference it indirectly by, for instance, using Sdk="Microsoft.NET.Sdk". I looked at one of the failing tests, and it was not an SDK-style project, and it didn't import that .targets file directly, which explains why it evaluated to "". You should be able to change the test asset to import it (and make sure you import it before Microsoft.TestPlatform.ImportAfter.targets).
> 
> Do you need this task in legacy .*proj files?

When you say legacy, do you mean non SDK-style projects?

**Comment by Forgind** [comment]
Yes

**Comment by mariam-abdulla** [comment]
> Yes

Then, we don't need the task there.

**Comment by Forgind** [comment]
> > Yes
> 
> Then, we don't need the task there.

Perfect, so we would just need to fix up the tests to make sure we only need to have that property defined in sdk-style test projects. Let me know if you need any help with that!

**Comment by mariam-abdulla** [comment]
> Perfect, so we would just need to fix up the tests to make sure we only need to have that property defined in sdk-style test projects. Let me know if you need any help with that!

Yes, please let me know what should be done in that case.

**Comment by MarcoRossignoli** [comment]
@Forgind do we have some ETA to merge this first part of @mariam-abdulla PR? We need it to move to the next and unblock the task.

**Review by MarcoRossignoli — APPROVED** [review]
LGTM and tests are green now @Forgind @baronfel are you good for the merge, can you review?

For now we went for "manual" resolution of `Microsoft.NET.Build.Tasks.dll` https://github.com/dotnet/sdk/pull/41142/files#diff-3f7b2597025eca7c8bdaa8632fab1a3a6894415a704c2b08c2b0760ac708ba39 to fix the test failure
And added the cross targeting part close to the other one https://github.com/dotnet/sdk/pull/41142/files#diff-177f9060fda1f47c96d4e08705bbaab76098d76df09f8911eeea06586db65fc6

**Comment by Forgind** [comment]
I haven't really reviewed this properly, so I'll leave it to @baronfel and @MiYanni (who is now domestic-cat) to decide when to merge.

**Review by baronfel — APPROVED** [review]
LGTM - the main thing was the Task location and that's been addressed. Thanks folks, excited to see how this evolves.

**Comment by baronfel** [comment]
@MarcoRossignoli / @mariam-abdulla please feel free to merge whenever you like - consider squashing though since many of the individual commits in this PR aren't meaningful to the end result.

**Comment by Evangelink** [comment]
Relates to #45927

**Pull request merged**
